### PR TITLE
Server routine: Listeners

### PR DESCRIPTION
**Pull request by EduPonz**

### CODE REVIEW

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp [hunk 1]
```diff
@@ -248,17 +252,59 @@ void PDPServer2::initializeParticipantProxyData(
...
+    logInfo(RTPS_PDP, "For RTPSParticipant: " << pdata->m_guid.guidPrefix);
+
+    const NetworkFactory& network = mp_RTPSParticipant->network_factory();
+    uint32_t endp = pdata->m_availableBuiltinEndpoints;
+    bool use_multicast_locators = !mp_RTPSParticipant->getAttributes().builtin.avoid_builtin_multicast ||
+            pdata->metatraffic_locators.unicast.empty();
+
+    // only SERVER and CLIENT participants will be received. All builtin must be there
+    uint32_t auxendp = endp & DISC_BUILTIN_ENDPOINT_PARTICIPANT_ANNOUNCER;
+    assert(0 != auxendp);
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:264]: On the one hand, assert will only get triggered in debug. On the other hand, the check should always be true, it's probably imposible to hit a situation where the sender is not a builtin

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp [hunk 2]
```diff
@@ -248,17 +252,59 @@ void PDPServer2::initializeParticipantProxyData(
...
+        temp_writer_data_.set_persistence_entity_id(c_EntityId_SPDPWriter);
+        temp_writer_data_.set_remote_locators(pdata->metatraffic_locators, network, use_multicast_locators);
+        temp_writer_data_.m_qos.m_reliability.kind = dds::RELIABLE_RELIABILITY_QOS;
+        temp_writer_data_.m_qos.m_durability.kind = dds::TRANSIENT_LOCAL_DURABILITY_QOS;
+        mp_PDPReader->matched_writer_add(temp_writer_data_);
+    }
+
+    // only SERVER and CLIENT participants will be received. All builtin must be there
+    auxendp = endp & DISC_BUILTIN_ENDPOINT_PARTICIPANT_DETECTOR;
+    assert(0 != auxendp);
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:281]: Same here

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp [hunk 3]
```diff
@@ -268,8 +314,22 @@ void PDPServer2::notifyAboveRemoteEndpoints(
 void PDPServer2::removeRemoteEndpoints(
         ParticipantProxyData* pdata)
 {
-    (void)pdata;
-    // TODO DISCOVERY SERVER VERSION 2
+    logInfo(RTPS_PDP, "For RTPSParticipant: " << pdata->m_guid);
+    uint32_t endp = pdata->m_availableBuiltinEndpoints;
+
+    assert(endp & DISC_BUILTIN_ENDPOINT_PARTICIPANT_ANNOUNCER);
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:320]: Same as in `assignRemoteEndpoints`

File: src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp [hunk 4]
```diff
@@ -268,8 +314,22 @@ void PDPServer2::notifyAboveRemoteEndpoints(
...
+
+    {
+        GUID_t writer_guid(pdata->m_guid.guidPrefix, c_EntityId_SPDPWriter);
+        mp_PDPReader->matched_writer_remove(writer_guid);
+    }
+
+    assert(endp & DISC_BUILTIN_ENDPOINT_PARTICIPANT_DETECTOR);
```
Comment (EduPonz) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer2.cpp:327]: Same as in `assignRemoteEndpoints`

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2704/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/95/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2501/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2715/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2711/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/100/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2508/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2722/)

**Pull request merged**